### PR TITLE
fs: use WTF-8 on Windows

### PR DESCRIPTION
**Pull request by seishun**
This allows working with filenames that are not well-formed UTF-16.

Since a `fs__wide_to_utf8` helper function already existed, I adapted it to be used across fs.c, added a similar function for the opposite direction and rewrote them to implement [WTF-8](https://simonsapin.github.io/wtf-8/).

### CODE REVIEW

**Comment by seishun** [comment]
It seems to be fully functional, but I submitted it as a draft because it's a significant change that needs to be discussed.

**Comment by seishun** [comment]
cc @rossj @bnoordhuis

**Comment by refack** [comment]
I'm not a fan of using WTF-8 as our internal representation, mostly from a K.I.S.S. aspect.
From the [WTF-8 spec](https://simonsapin.github.io/wtf-8/#intended-audience):
* image

so if we don't ***need*** to use WTF-8 we shouldn't, and assume other systems are self-contained WRT to WTF-8, and hence leave the system in a consistent state WRT UTF-8.

* image

i.e. there is no way to create a broken fs entry with the libuv API. The test submitted uses a direct WIN32 call:
https://github.com/libuv/libuv/blob/894384265298f41471ce9b269ed9d524424b18fa/test/test-fs.c#L3801

but, nice catch @seishun!

**Comment by rossj** [comment]
@refack -
There's some more discussion over at issue #2048, but the short of it is that currently there may exist files on a system that `libuv` can't interact with at all, except for being able to get a mangled string representation of the filename from a dir listing operation. I've run into this issue on a production system a couple times, where my Node program was unable to clean and remove a directory due to it containing ill-formed UTF-16 files.

As for the test, I believe that it is based on my example in issue #2048. This PR does allow creating & interacting with these problem files by passing in WTF-8 input, so the test case could likely be updated to not use the WIN32 call.

@seishun -
First, thank you for taking the initiative and working on this!

I only have one code note - It looks like the code, primarily `fs__decode_wtf8_char`, assumes that the incoming filenames are at least well-formed WTF-8. Perhaps `fs__decode_wtf8_char` should have a final fall-through `else` and some extra assertions that the extra bytes being consumed are present and proper.

**Comment by refack** [comment]
> there may exist files on a system that `libuv` can't interact with at all

Just as a data point, I just tested and Python (2 & 3) fail in a similar manor...

Also I would assume there are other ways to brute-force generate invalid fs entries, or several of the supported fses.
I'm not sure `libuv` should have fsck like powers, to fix those issues ¯\\_(ツ)_/¯

P.S. even Windows Explorer can encounter fs entries it can't handle
https://github.com/Microsoft/WSL/issues/3853
or the ubiquitous
image

**Comment by rossj** [comment]
In my view, these files we're talking about are not invalid fs entries. They are valid, they just do not represent a valid UTF-16 string, which is allowed (albeit unusual). In my testing, such files do not cause any issue for Microsoft programs such as Windows Explorer, Office, Outlook.

Per [Microsoft](https://docs.microsoft.com/en-us/windows/desktop/FileIO/naming-a-file) "There is no need to perform any Unicode normalization on path and file name strings for use by the Windows file I/O API functions because the file system treats path and file names as an opaque sequence of WCHARs."

I do think it's a perfectly valid stance that these invalid UTF-16 files are rare enough that they don't warrant support. In that case, I think the documentation should be updated to reflect that on Windows, any files or directories with non-UTF-16 names cannot be interacted with, and returned paths and filenames from listing operations may not accurately or uniquely identify the underlying file.

**Comment by seishun** [comment]
>This PR does allow creating & interacting with these problem files by passing in WTF-8 input, so the test case could likely be updated to not use the WIN32 call.

I intentionally didn't use libuv API because the purpose of this PR is to work with _existing_ non-UTF-16 filenames. More specifically:
* The libuv call might accidentally "normalize" the filename, and the test would still pass.
* We might want to disallow creating _new_ non-UTF-16 filenames.
>I only have one code note - It looks like the code, primarily `fs__decode_wtf8_char`, assumes that the incoming filenames are at least well-formed WTF-8.

I followed the [spec](https://simonsapin.github.io/wtf-8/#decoding-wtf-8): _Since WTF-8 must not be used for interchange (see Intended audience), this algorithm is deliberately not defined for arbitrary byte sequences. It is only defined for byte sequences known to be well-formed in WTF-8, such as sequences encoded from code points, converted from UTF-16, or concatenated from sequences themselves well-formed in WTF-8._

The libuv embedder must make sure the input is well-formed WTF-8, otherwise there is a risk they will pass filenames directly from the end user. But there are assertions to validate the input in a debug build.

**Comment by StefanKarpinski** [comment]
I want to express strong support form Julia for this change. The problem is that libuv needs to decide what to do if the file system gives it invalid UTF-16. The reasonable options are:

1. Choke and throw an error;
2. Represent it as WTF-8.

WTF-8 is a way of representing arbitrary valid/invalid UTF-16 data as similarly invalid but well-formed "generalized UTF-8" such that if you transcode it back to UTF-16 you get the exact same UTF-16 data back. Moreover, for most operations like "splitting on `/`" or other common Unicode-oblivious C operations on file names, WTF-8 will be handled correctly and the result will be sane, having invalid data in the end only where you'd expect it.

Note that on systems where file system operations natively return UTF-8 data, it is already completely possible to get invalid UTF-8 from the file system, so this is no worse than that. Invalid UTF-8 already has to be expected and handled by applications using these libuv APIs that are supposed to be robust. Since WTF-8 is at least well-formed generalized UTF-8, it is substantially less likely to cause any problems than the arbitrarily malformed invalid UTF-8 that can be returned natively.

**Comment by StefanKarpinski** [comment]
> so if we don't need to use WTF-8 we shouldn't

I would argue that libuv has no choice but using WTF-8 here since throwing an error when trying to work with file names that the application has no control over is not ok.

**Comment by StefanKarpinski** [comment]
Bump. What's required for forward progress here?

**Comment by bnoordhuis** [comment]
@libuv/windows This is within your remit.

**Comment by vtjnash** [comment]
Draft implies there's still work to be done, but since OP says it's fully functional sounds like this should actually be marked "ready for review" (it's already got the RFC tag)?

Anyways, +1 from me. It seems like this fits in well with JuliaLang's general approach to byte-string handling/conversion (as expressed above by StefanKarpinski in more detail), leaving it up to the high-level display layer to deal with representation details  and treat the bytes opaquely at the low level like on unix.

**Comment by bnoordhuis** [comment]
@vtjnash What about backwards compatibility? See discussion in #2048.

**Comment by StefanKarpinski** [comment]
Regarding compatibility, are you talking about this comment?

> For libuv v1.x, it might well be that we decide to do nothing because of the risk of regressions in downstream projects.

The risk to downstream projects is minimal, and indeed by my analysis any cross-platform usage of libuv that is currently correct would still be correct. Any program that could be negatively affected by this change either doesn't ever run on UNIX or is already broken in a far more dangerous way on UNIX.

To expand on that, this can only cause problems if a project using libuv is relying on always getting valid UTF-8 from libuv's API. However, that is not a documented guarantee and is not true on UNIX platforms: on UNIX the file system can and will return arbitrarily malformed UTF-8 data. On Windows, even with this change, the API will always return WTF-8, which consists of well-formed UTF-8 code points—i.e. you can decode it into numeric code points—it's just that some of them are surrogates, which are "not allowed" as code points. Returning WTF-8 is a *much* smaller issue than returning totally arbitrary data, since most code that expects UTF-8 will work just fine on WTF-8 unless it explicitly checks for code points validity, in which case it almost certainly will handle that case.

In summary:

* This could only potentially affect programs using libuv on Windows systems which would previously have failed with a file system error.
* Any program which is currently correct on both Windows and UNIX systems will not break since it must already handle getting arbitrary invalid UTF-8 data on UNIX.
* Any bug due to mishandling of invalid UTF-8 returned on Windows would be worse on UNIX since there you can get malformed UTF-8 rather than merely invalid UTF-8.
* Most code which naively expects valid UTF-8 will work correctly when fed WTF-8 since it has the correct structure. It is more work to check that the decoded code points are valid, and software which checks for invalid code points is already checking for invalid UTF-8 and therefore will not break.

**Comment by StefanKarpinski** [comment]
Or to put it another way, each downstream program falls into one of three categories in terms of how it interacts with the supposedly-UTF-8 data returned by libuv filesystem functions:

1. Doesn't care if it's valid or not.
2. Assumes validity.
3. Checks validity.

Case 1 would be unaffected. Programs in case 2 would also be unaffected since they would decode the WTF-8 data just fine and then not notice the fact that the code points happen to be for surrogates because they're not checking for validity. Programs in case 3 would go from getting an invalid Unicode error from libuv to throwing their own invalid Unicode error.

**Comment by StefanKarpinski** [comment]
Bump, can we get some feedback or better yet a decision on this?

**Comment by bnoordhuis** [comment]
@libuv/windows This needs more input from you. @vtjnash already +1'd it.

**Comment by StefanKarpinski** [comment]
Any chance of ever getting input from @libuv/windows? My general approach when contributors ignore issues is to go ahead with the change. Nothing solicits feedback like changing something.

**Comment by saghul** [comment]
@libuv/windows ping?

File: src/win/fs.c [hunk 1]
```diff
@@ -142,6 +142,57 @@ void uv_fs_init(void) {
 }
 
 
+static unsigned int fs__decode_wtf8_char(const char** input) {
+  unsigned int code_point;
+  unsigned char b = **input, b2, b3, b4;
+  if (b >= 0x00 && b <= 0x7F) {
+    code_point = b;
+  } else if (b >= 0xC2 && b <= 0xDF) {
+    assert((*input)[1]);
```
Comment (vtjnash) [inline src/win/fs.c]: This input is user text, so it shouldn't be an assert here. Let's report this back as `STATUS_INVALID_PARAMETER` / `UV_ENOTDIR` if the value `b2` doesn't have the correct bits set.

File: src/win/fs.c [hunk 2]
```diff
@@ -142,6 +142,57 @@ void uv_fs_init(void) {
 }
 
 
+static unsigned int fs__decode_wtf8_char(const char** input) {
+  unsigned int code_point;
+  unsigned char b = **input, b2, b3, b4;
+  if (b >= 0x00 && b <= 0x7F) {
```
Comment (vtjnash) [inline src/win/fs.c]:
```suggestion
  if (b <= 0x7F) {
```
(b is unsigned, so this would generate a compiler warning)

File: src/win/fs.c [hunk 3]
```diff
@@ -142,6 +142,57 @@ void uv_fs_init(void) {
...
+    code_point = ((b & 0x0F) << 12) + ((b2 & 0x3F) << 6) + (b3 & 0x3F);
+  } else if (b >= 0xF0 && b <= 0xF0) {
+    assert((*input)[1] && (*input)[2] && (*input)[3]);
+    b2 = *++*input;
+    b3 = *++*input;
+    b4 = *++*input;
+    code_point = ((b & 0x07) << 18) + ((b2 & 0x3F) << 12) +
+      ((b3 & 0x3F) << 6) + (b4 & 0x3F);
+  }
+  assert(code_point <= 1114111);
```
Comment (vtjnash) [inline src/win/fs.c]:
```suggestion
  assert(code_point <= 1114111); /* 0x10ffff */
```
(though, again, we should make this return an error, not assert it)

File: src/win/fs.c [hunk 4]
```diff
@@ -142,6 +142,57 @@ void uv_fs_init(void) {
...
+    b3 = *++*input;
+    code_point = ((b & 0x0F) << 12) + ((b2 & 0x3F) << 6) + (b3 & 0x3F);
+  } else if (b >= 0xF0 && b <= 0xF0) {
+    assert((*input)[1] && (*input)[2] && (*input)[3]);
+    b2 = *++*input;
+    b3 = *++*input;
+    b4 = *++*input;
+    code_point = ((b & 0x07) << 18) + ((b2 & 0x3F) << 12) +
+      ((b3 & 0x3F) << 6) + (b4 & 0x3F);
+  }
```
Comment (vtjnash) [inline src/win/fs.c]:
```suggestion
  } else { /* b >= 0x80 && b <= 0xC2 */
      what happens here? (I think it's malformed utf8/wtf8, but it still should be handled)
  }
```

**Comment by seishun** [comment]
@vtjnash Thank you for the comments. Some of them relate to arbitrary user input, which I addressed in the second half of https://github.com/libuv/libuv/pull/2192#issuecomment-464457965. What do you think?

File: src/win/fs.c (hunk 4 above)
Comment (StefanKarpinski) [reply src/win/fs.c]: Unlike the other direction where every possible valid/invalid UTF-16 sequence can be encoded as WTF-8 and a unique and roundtrippable way, there's no way to encode malformed (as opposed to invalid) WTF-8 as UTF-16. Malformed WTF-8 like this would have to come from the user side, wouldn't it? Two possible options:

1. Replace malformed sequences with replacement characters (U+FFFD); according to the Unicode standard;
2. Throw an error since invalid data was coming from user and on Windows there's no way to respect it.

If you choose option 1, the Unicode standard says that you should replace each well-formed but too high utf8 byte sequence with a single replacement character and emit one replacement character for each malformed byte sequence as soon as it includes a byte that precludes the sequence from being well-formed. However, since the user side, as opposed to the system side, is passing in bad data that cannot be respected here, raising an error seems reasonable.

**Comment by StefanKarpinski** [comment]
Reopen, stalebot.

File: src/win/fs.c (hunk 4 above)
Comment (StefanKarpinski) [reply src/win/fs.c]: I think the best course of action here is to throw an error because unlike when the file system has invalid UTF-16, which is beyond the user's control and is allowed by most file systems (both Windows and UNIX allow arbitrary invalid data in file names), this is within the user's control and there is no reasonable way to handle it on Windows: there is simply no way to take a _malformed_ (as opposed to invalid but still well-formed WTF-8) UTF-8 string and ask for a path with that name on a UTF-16 file system—this is simply an impossibility.

Comment (seishun) [reply src/win/fs.c]: @StefanKarpinski see https://github.com/libuv/libuv/pull/2192#issuecomment-464457965. The string won't come directly from the user (who shouldn't know anything about WTF-8) but from the embedder, so we can expect valid input.

Comment (StefanKarpinski) [reply src/win/fs.c]: All the more reason to error of that assumption fails :+1:

Comment (seishun) [reply src/win/fs.c]: An error would make sense if it can be handled. In this case, that assumption failing indicates a problem in the embedder's code, so an assert is more appropriate.

**Comment by StefanKarpinski** [comment]
Reopen, stalebot. It's super fun doing this every 32 days.

**Comment by StefanKarpinski** [comment]
Is it already that time again? How fast yet slow the time has gone by. I've decided that I no longer care enough to continue to keep this open.

**Comment by vtjnash** [comment]
It is somewhat stale though, in that I'm waiting for my review comments to be addressed so that this can be merged.

File: src/win/fs.c (hunk 4 above)
Comment (vtjnash) [reply src/win/fs.c:166]: libuv, as a design principle, does not abort when given bad input that could be detected, and therefore handled (thus generally only for passing in memory of the wrong type, since we can't protect the user against malloc/free mistakes)

Comment (seishun) [reply src/win/fs.c:166]: Let me explain my point of view in more detail.

WTF-8 is [intended to be used internally in self-contained systems](https://simonsapin.github.io/wtf-8/#intended-audience). That means external input (input originating outside the system) must not be WTF-8. Consequently, the program that uses libuv must not pass external input to libuv as-is, since it might be WTF-8. In other words, the input would either be produced by the program or come from libuv itself. So any decoding error would indicate a bug in the program or in libuv. Given that, how could the program meaningfully handle the error?

Comment (vtjnash) [reply src/win/fs.c:166]: I understand your point of view, and I'm telling you that libuv is not written with that philosophy of punishing the user with abort, but is written more like libc or the kernel, which are expected to return an error code and not kill your program. The use of `assert` (or abort) is for internal logic errors within libuv or egregious errors with memory (since neither can be safely and reliably detected). When errors can be returned to the embedding program safely (such as here), they can be reported as EINVAL.

Comment (seishun) [reply src/win/fs.c:166]: Well, I'm not comfortable with deviating further from the spec. We're already pushing it by exposing WTF-8 outside libuv, and I suspect that tolerating invalid input would make it more likely to be exposed to end users.

Perhaps WTF-8 isn't the right solution for libuv.

Comment (vtjnash) [reply src/win/fs.c:166]: I'm not asking you to deviate from the spec in any way. I'm simply asking that it follows the existing convention in libuv and return EINVAL as an error code when encountering input that it cannot handle. (and probably a significant improvement over the current situation, which is either to silently drop them or replaces them with U+FFFD depending on the OS version)

**Comment by JeffBezanson** [comment]
Bump

File: src/win/fs.c (hunk 4 above)
Comment (seishun) [reply src/win/fs.c:166]: See https://simonsapin.github.io/wtf-8/#decoding-wtf-8, for instance. 

Can we get another opinion here? I don't insist that all specs be followed to the letter, but if we're having a disagreement about a fundamental design decision in WTF-8, then it's an indication that libuv might not be the intended audience.

Comment (vtjnash) [reply src/win/fs.c:166]: Sure, but currently your implementation also is a violation of 6.3 (“This conversion never fails” or asserts, even on ill formed input). Elsewhere, the document gives 3 options for handling ill formed WTF-8 (see the section where it defines it): Reject it, ignore it, or replace with the replacement character. We’ve previously done both 2 and 3. Any of the three are likely acceptable behaviors here, so I lean towards 1. 

All you aren’t allowed to do is abort on verifiable user input errors, as that violates both the spec you just linked and the general design principles of libuv in its role as a systems library.

Comment (StefanKarpinski) [reply src/win/fs.c:166]: This is pretty simple and has nothing to do with WTF-8:  it is unacceptable for a library like libuv (or any C library for that matter) to call `abort`. Ever. C libraries _must_ indicate errors with error codes. That's all @vtjnash is asking you to do here. If you won't do that, then this PR simply cannot be merged as is. In that case, someone will have to take this PR and modify it to return error codes instead of calling `abort`.

**Comment by StefanKarpinski** [comment]
I made a PR that just applies the changes that @vtjnash requested here: https://github.com/libuv/libuv/pull/2970.

File: src/win/fs.c (hunk 4 above)
Comment (seishun) [reply src/win/fs.c:166]: >Sure, but currently your implementation also is a violation of 6.3 (“This conversion never fails” or asserts, even on ill formed input).

That's clearly an oversight given https://github.com/SimonSapin/wtf-8/commit/c6a271dc8090f6534127abfb6ad651ef1aff9023. I've submitted an issue here: https://github.com/SimonSapin/wtf-8/issues/13

>Elsewhere, the document gives 3 options for handling ill formed WTF-8 (see the section where it defines it): Reject it, ignore it, or replace with the replacement character.

Could you link to that section?

>it is unacceptable for a C libraries to call abort

I agree with that if we consider libuv's API the system's boundary. But in that case libuv is not the intended audience for WTF-8 ("Any WTF-8 data must be converted to a Unicode encoding at the system’s boundary before being emitted.").

**Comment by vtjnash** [comment]
Replaced by #2970
